### ISSUE TITLE
Check/sanitise the `expire_logs_after` value

### ISSUE THREAD

**Issue by petecooper** [textpattern/textpattern#1600]
The `expire_logs_after` input accepts various values that aren't valid numbers of days:

![Screenshot 2020-12-05 at 18 25 08](https://user-images.githubusercontent.com/226578/101260493-4de4ec80-3728-11eb-8bfc-c6ebf976e832.png)
![Screenshot 2020-12-05 at 18 25 19](https://user-images.githubusercontent.com/226578/101260495-4f161980-3728-11eb-8499-6d9ec21964fc.png)
![Screenshot 2020-12-05 at 18 25 51](https://user-images.githubusercontent.com/226578/101260496-4faeb000-3728-11eb-8d40-7986ccb1c2cf.png)

Is there a sane way we can check this value is a) numeric and b) ≥ 0?

**Comment by petecooper** [comment]
Also affects `rss_how_many` and `max_url_len`.

**Comment by Bloke** [comment]
Good catch. We have the vendors/Textpattern/Validator class that we employ elsewhere in the workflow, so we might as well use it on prefs too. Leave it with me.

**Comment by Bloke** [comment]
I have a patch ready that will test just this one violation for now (will do the others soon). But I have a question: if it fails a constraint during prefs save, what do we want to happen? Warn and tell you which one(s) failed validation, and then:

A: Carry on saving all the values anyway (we've told you which were borked: up to you to fix it).

B: Save all the values bar the ones that failed (which means the erroneous values revert to their previous values so you don't know what the new value you were trying to save was).

C: Save none of the prefs until you fix the issue(s).

At the moment it does A, which means the bad value gets stored anyway.

I'm kinda leaning towards doing B, although it might be annoying. What if, in future, we validated the email address in 'Send login details from this email address' and found it wasn't valid due to a typo in the domain name? You'd be warned and the field would be reverted to its previous value rather than leaving the (bad) address for you to fix the one character you mistyped. One way leaves bad prefs values available in the database, the other doesn't but inconveniences you.

Option C will require some fairly heavy code refactorisation as we won't be able to save the prefs as we go. We'd either have to batch them all up and do them in one go once everything has checked out, or we'd have to wrap the lot in a transaction and back out if anything failed validation. But do we want to penalise _other_ prefs by preventing them being saved just because one of the prefs was out of spec?

None of the options are ideal, imo. But leaving it without validation is also incorrect.

Are there are any other saving options we could explore to make this more ideal?

**Comment by phiw13** [comment]
Can’t you do the same as e.g. the date & time fields on the Write panel: `inputmode="numeric" pattern="xxx"`

Safari lets you fill-in the field but doesn’t let your save; Firefox, after filling with a non-matching pattern (letters, out-of-range numbers) immediately trigger the `:-moz-ui-invalid` pseudo-class and then refuses to save. Both browsers pop up an alert bubble on the (first) offending field. Safari doesn’t have an equivalent pseudo-class yet. 

Chromium based browsers untested as not installed on this MBA.

The `inputmode` is mostly of value for iOS/Android devices.

**Comment by petecooper** [comment]
@Bloke – given relative readiness of patch to test, I've tacked `4.8.5` milestone, please bump to `4.9` if you prefer.

**Comment by Bloke** [comment]
@phiw13 Yes, kind of. Two things at present make in-browser validation harder.

1) There's no validation rules provision in the database against each pref. If there were, when we loop over the fields to build the panel content we could add any constraints there. Without this, it's a lot of extra hard-coded "what field is this? Oh right, yeah, add this to the markup" stuff. And there's already a lot of that.

2) With the in-browser flicking of panels via the TOC, what happens if you alter a pref, keep focus in the box, click away to another subpane, hiding the one you were working on, edit another pref, then click Save? I suspect the browser won't flick to the correct subpane to show where the error is. And if there's more than one subpane containing validation errors, what then?

There's scope for auto saving each pref and removing the save button entirely, but I'm not a massive fan of that UX, especially since some settings you want to apply in tandem.

Alternatively, the Save could only store the current pane's settings. Again, not a huge fan of this.

Or, if you try to flick to another area of the panel, we could perhaps trigger a sweep of all the visible prefs and prevent you leaving the subpane until the validation is fixed. @bloatware might have an idea on whether this is viable to ask the browser to check on click.

Either way, there's a lot to think about if we go with in-browser validation. Plus, we still need to do server-side validation: it's good practice in case anyone bypasses the restrictions by editing the DOM in the inspector, so we'd have to get the server-side workflow right anyway.

**Comment by phiw13** [comment]
> With the in-browser flicking of panels via the TOC, […]

Hmm… I was under the impression that when toggling to another view (sub panel) the un-saved changes in the older view where discarded. If that is not the case, then indeed, that complicates the issue a bit (understatement?).

**Comment by petecooper** [comment]
Bumped to 4.9 given complexities, should not be a blocker for 4.8.5.

**Comment by philwareham** [comment]
Can any of this be alleviated by the new widgets UI that Stef has built but we haven't applied yet? Certainly the `pattern` and `inputmode` attributes could be harnessed.

As for validation, we could introduce some AJAX-like validation with some work, maybe?

Short of that, even if we didn't I think it's better to fail the save if invalid input values exist rather than save erroneous values. That's better then the current UX at least. It'd then be up to the user to go through the form and find their incorrect values. Not perfect but at start to build upon.

Definitely a 4.9.x issue though, not 4.8.5 blocker.

**Comment by Bloke** [comment]
Yeah, this might be better left for 4.9 since the UI library can handle this. We still need some database-level indication of expected format for each field (blank if we don't care) as that can then automagically build the inputs from the database content for each pref and, crucially, use the value there to validate both browser-side and server-side entries.

> it's better to fail the save if invalid input values exist rather than save erroneous values.

I tend to agree. There's no obvious (to me) save route that could list which fields have failed browser-side validation, but at least if you get past that and do manage to try and save something bogus, the returned failure message will list the fields that have violated their requirements.

**Comment by petecooper** [comment]
Tagging https://github.com/textpattern/textpattern/issues/355 as a similar sanitising issue that would could address around the same time.
